Announce opam 2.2.0 (#103)
ocaml/platform-blog: opam-2-2-0.md

  "ocaml"
  "ocp-indent" {with-dev-setup}
]
build: [make]
install: [make "install"]
post-messages:
[ "Thanks for installing the package"
  "as well as its development setup. It will help with your future contributions" {with-dev-setup} ]
```

### Major change: opam pin --recursive

When pinning a package using `opam pin`, opam looks for opam files in the root directory only.
With recursive pinning, you can now instruct opam to look for `.opam` files in
subdirectories as well, while maintaining the correct relationship between the `.opam`
files and the package root for versioning and build purposes.

Recursive pinning is enabled by the following options to `opam pin` and `opam install`:
 * With `--recursive`, opam will look for `.opam` files recursively in all subdirectories.
 * With `--subpath <path>`, opam will only look for `.opam` files in the subdirectory `<path>`.

The two options can be combined: for instance, if your opam packages are stored
as a deep hierarchy in the `mylib` subdirectory of your project you can try
`opam pin . --recursive --subpath mylib`.

These options are useful when dealing with a large monorepo-type repository with many
opam libraries spread about.

> 
> ### Major change: --formula
> 
> It is now possible to leverage the full expressivity of package dependency
> formulas from the command line during switch creation and package operations.
> 
> For example, it's now possible to create a switch using a formula:
> ```
> opam switch create 4.14 --formula '"ocaml-base-compiler" {>= "4.14.0" & < "5.0.0~~"}'
> ```
> this will let you create self upgrading opam switches. In the given example above
> whenever a new version of OCaml 4.14 is available, opam will ask to upgrade.

Author: ```suggestion
```

### New Options

* `opam switch -`, inspired by `git switch -`, makes opam switch back to the previously
  selected global switch.

* `opam pin --current` fixes a package to its current state (disabling pending
  reinstallations or removals from the repository). The installed package will
  be pinned to its current installed state, i.e. the pinned opam file is the 
  one installed.

* `opam pin remove --all` removes all the pinned packages from a switch.

* `opam exec --no-switch` removes the opam environment when running a command.

* `opam clean --untracked` removes untracked files interactively remaining
  from previous packages removal.

* `opam admin add-constraint <cst> --packages pkg1,pkg2,pkg3` applies the given constraint
  to a given set of packages

> * `opam list --base` has been renamed into `--invariant`, reflecting the fact that since opam 2.1 the "base" packages of a switch are instead expressed using a switch invariant.

Author: ```suggestion
* `opam list --base` has been renamed into `--invariant`, reflecting the fact that since opam 2.1 the "base" packages of a switch are instead expressed using a switch invariant.

* `opam install --formula <formula>` installs a formula instead of a list of packages. This can be useful if you would like to install one package or another one. For example `opam install --formula '"extlib" |"extlib-compat"'` will install either `extlib` or `extlib-compat` depending on what's best for the current switch.
```

- [the release note for 2.2.0~beta3](https://github.com/ocaml/opam/releases/tag/2.2.0-beta3)
- [the release note for 2.2.0~rc1](https://github.com/ocaml/opam/releases/tag/2.2.0-rc1)
This release also includes PRs improving the documentation and improving
and extending the tests.

## Try it!

In case you plan a possible rollback, you may want to first backup your
`~/.opam` or `$env:LOCALAPPDATA\opam` directory.

The upgrade instructions are unchanged:

1. Either from binaries: run

For Unix systems
```
bash -c "sh <(curl -fsSL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh) --version 2.2.0"
```
or from PowerShell for Windows systems
```
Invoke-Expression "& { $(Invoke-RestMethod https://raw.githubusercontent.com/ocaml/opam/master/shell/install.ps1) }"
```
or download manually from [the Github "Releases" page](https://github.com/ocaml/opam/releases/tag/2.2.0) to your PATH.

2. Or from source, manually: see the instructions in the [README](https://github.com/ocaml/opam/tree/2.2.0#compiling-this-repo).

You should then run:
```
opam init --reinit -ni
```

Please report any issues to [the bug-tracker](https://github.com/ocaml/opam/issues).

> We hope you will enjoy the new features of opam 2.2!

Reviewer: ```suggestion
We hope you will enjoy the new features of opam 2.2! 📯
```

Author: XDD

title: "opam 2.2.0 release"
authors: [
  "Raja Boujbel - OCamlPro" {"mailto:raja.boujbel(à)ocamlpro.com"}
  "Kate Deplaix - Ahrefs" {"mailto:kit-ty-kate(à)outlook.com"}
  "David Allsopp - Tarides" {"mailto:david(à)tarides.com"}
]
date: "2024-07-01"
--BODY--

_Feedback on this post is welcomed on [Discuss](https://discuss.ocaml.org/t/ann-opam-2-2-0/XXXXX)!_

> We are very excited to announce the release of opam 2.2.0.
> 
> We've squashed a couple of bugs since the release candidate and for those interested you can read more about in the [release note](https://github.com/ocaml/opam/releases/tag/2.2.0).

Reviewer 2: I don't think the bug squashing and release note matter so much for the release itself:

```suggestion
We are very pleased to announce the release of opam 2.2.0, and encourage all users to upgrade. Please read on or visit the [release page](https://github.com/ocaml/opam/releases/tag/2.2.0) for installation and upgrade instructions.
```

Reviewer: hmm... there is no install/upgrade instruction in the release note :)

Reviewer 2: I know, but I figured that if someone was then confused by the binaries, the release note guides them back to the blog post :)

Reviewer: why not link to the install instruction at the top of the article?
Anyway, we can also add upgrade instruction in the release note, or add a link to the good section, regardless of this sentence.

```

Reverse-dependencies can also be displayed using the new `opam why` command.
This is useful to examine how dependency versions get constrained.
```
$ opam why cmdliner
cmdliner.1.2.0
├── (>= 1.1.0) b0.0.0.5
│   └── (= 0.0.5) odig.0.0.9
├── (>= 1.1.0) ocp-browser.1.3.4
├── (>= 1.0.0) ocp-indent.1.8.1
│   └── (>= 1.4.2) ocp-index.1.3.4
│       └── (= version) ocp-browser.1.3.4 [*]
├── (>= 1.1.0) ocp-index.1.3.4 [*]
├── (>= 1.1.0) odig.0.0.9 [*]
├── (>= 1.0.0) odoc.2.2.0
│   └── (>= 2.0.0) odig.0.0.9 [*]
├── (>= 1.1.0) opam-client.2.2.0~alpha
│   ├── (= version) opam.2.2.0~alpha
│   └── (= version) opam-devel.2.2.0~alpha
├── (>= 1.1.0) opam-devel.2.2.0~alpha [*]
├── (>= 0.9.8) opam-installer.2.2.0~alpha
└── user-setup.0.7
```

> Special thanks to [@cannorin](https://github.com/cannorin) for contributing this feature.

### Major change: with-dev-setup

There is now a way for a project maintainer to share their project development
tools: the `with-dev-setup` dependency flag. It is used in the same way as
`with-doc` and `with-test`: by adding a `{with-dev-setup}` filter after a
dependency. It will be ignored when installing normally, but it's pulled in when the

> package is explicitely installed with the `--with-dev-setup` flag specified on

Reviewer 2: ```suggestion
package is explicitly installed with the `--with-dev-setup` flag specified on
```

the command line.

For example
```
opam-version: "2.0"
depends: [
  "ocaml"
  "ocp-indent" {with-dev-setup}
]
build: [make]
install: [make "install"]
post-messages:
[ "Thanks for installing the package"

>   "as well as its developement setup. It will help for your futur contributions" {with-dev-setup} ]

Reviewer 2: ```suggestion
  "as well as its development setup. It will help with your future contributions" {with-dev-setup} ]
```

```

### Major change: opam pin --recursive

When pinning a package using `opam pin`, opam looks for opam files in the root directory only.

> With recursive pinning, you can now instruct opam to also look for `.opam` files in
> subdirectories, while maintaining the correct relationship between the `.opam`

Reviewer 2: I'm not usually pedantic about split infinitives, but I think this one can be avoided:
```suggestion
With recursive pinning, you can now instruct opam to look for `.opam` files in
subdirectories as well, while maintaining the correct relationship between the `.opam`
```

files and the package root for versioning and build purposes.

> Recursive pinning is enabled the following options to `opam pin` and `opam install`:

Reviewer 2: ```suggestion
Recursive pinning is enabled by the following options to `opam pin` and `opam install`:
```

 * With `--recursive`, opam will look for `.opam` files recursively in all subdirectories.
 * With `--subpath <path>`, opam will only look for `.opam` files in the subdirectory `<path>`.

The two options can be combined: for instance, if your opam packages are stored
as a deep hierarchy in the `mylib` subdirectory of your project you can try
`opam pin . --recursive --subpath mylib`.

These options are useful when dealing with a large monorepo-type repository with many
opam libraries spread about.

### Major change: --formula

It is now possible to leverage the full expressivity of package dependency
formulas from the command line during switch creation and package operations.

For example, it's now possible to create a switch using a formula:
```
opam switch create 4.14 --formula '"ocaml-base-compiler" {>= "4.14.0" & < "5.0.0~~"}'
```
this will let you create self upgrading opam switches. In the given example above

> whenever a new version of OCaml 4.14 is available, opam will ask to upgrade.

Reviewer 2: A TODO note, but `opam switch create --formula` is a 2.1 feature - the new feature is being able to do that with `opam install`. It's a side-effect of fixing bugs IIRC, and it may be worth mentioning that.

Author: ah, my bad. In that case i'm not sure it's worth mentioning in its own section. I'll have a look into moving it into the "New options" section

### New Options

> * `opam switch -`, inspired from `git switch -`, makes opam switch back to the previously

Reviewer 2: ```suggestion
* `opam switch -`, inspired by `git switch -`, makes opam switch back to the previously
```

  selected global switch.

* `opam pin --current` fixes a package to its current state (disabling pending
  reinstallations or removals from the repository). The installed package will
  be pinned with the opam file that is stored in opam internal state, the one

>   that is currently installed.

Reviewer 2: TODO: this sentence is unclear to me

* `opam pin remove --all` removes all the pinned packages from a switch.

> * `opam exec --no-switch` removes the opam environment when running a command.

Reviewer 2: TODO: use case

Reviewer: ```suggestion
* `opam exec --no-switch` removes the opam environment when running a command.
  It is useful when you want to launch a command without opam environment changes.
```

* `opam clean --untracked` removes untracked files interactively remaining
  from previous packages removal.

* `opam admin add-constraint <cst> --packages pkg1,pkg2,pkg3` applies the given constraint
  to a given set of packages

> * `--base` has been renamed into `--invariant`.

Reviewer 2: ```suggestion
* `opam list --base` has been renamed into `--invariant`, reflecting the fact that since opam 2.1 the "base" packages of a switch are instead expressed using a switch invariant.
```

Reviewer 2: TODO: the semantics are slightly different as well, and that's probably worth mentioning?

### Miscellaneous changes

* The UI now displays a status when extracting an archive or reloading a repository

> * Fix a rounding error when displaying the timestamp in debug mode

Reviewer 2: I don't think we need to mention this one 🙈
```suggestion
```

> * Fix several edge cases related to environment reverting

Reviewer 2: ```suggestion
* Overhauled the implementation of `opam env`, fixing many corner cases for environment updates and making the reverting of package environment variables precise. As a result, using `setenv` in an opam file no longer triggers a lint warning.
```

* Fix parsing pre-opam 2.1.4 switch import files containing extra-files
* Add a new `sys-ocaml-system` default global eval variable
* Hijack the `"%{var?string-if-true:string-if-false-or-undefined}%"` syntax to
  support extending the variables of packages with `+` in their name
  (`conf-c++` and `conf-g++` already exist) using `"%{?pgkname:var:}%"`
* Fix issues when using fish as shell
* Sandbox: Mark the user temporary directory
  (as returned by `getconf DARWIN_USER_TEMP_DIR`) as writable when TMPDIR
  is not defined on macOS
* Add Warning 69: Warn for new syntax when package name in variable in string
  interpolation contains several '+' (this is related to the "hijack" item above)

> * Add support for Wolfi OS, treat it like Alpine family as it also uses apk

Reviewer 2: ```suggestion
* Add support for Wolfi OS, treating it like Alpine family as it also uses apk
```

> * Move last-env `OPAM_LAST_ENV` files outside the switch to be in `$OPAMROOT`

Reviewer 2: ```suggestion
```
(part of the overhaul above)

> * Remove the unused/untested lib-pkg Makefile targets

Reviewer 2: Not sure this is needed (although it was used and it was tested prior to its removal!!)
```suggestion
```

> * Remove the dependency towards cppo

Reviewer 2: ??
```suggestion
```

  (`conf-c++` and `conf-g++` already exist) using `"%{?pgkname:var:}%"`
* Fix issues when using fish as shell
* Sandbox: Mark the user temporary directory
  (as returned by `getconf DARWIN_USER_TEMP_DIR`) as writable when TMPDIR
  is not defined on macOS
* Add Warning 69: Warn for new syntax when package name in variable in string
  interpolation contains several '+' (this is related to the "hijack" item above)
* Add support for Wolfi OS, treating it like Alpine family as it also uses apk
* Sandbox: `/tmp` is now writable again, restoring POSIX compliance
* Add a new `opam admin: new add-extrafiles` command to add/check/update the `extra-files:` field according to the files present in the `files/` directory
* Add a new `opam lint -W @1..9` syntax to allow marking a set of warnings as errors
* Fix bugs in the handling of the `OPAMCURL`, `OPAMFETCH` and `OPAMVERBOSE` environment variables
* Fix bugs in the handling of the `--assume-built` argument
* Software Heritage fallbacks is now supported, but is disabled-by-default for now. For more information you can read one of our [previous blog post](https://opam.ocaml.org/blog/opam-2-2-0-alpha/#Software-Heritage-Binding)

> And many other general and performance improvements were made and bugs were fixed.

Reviewer: ```suggestion
And many other general and performance improvements were made and bugs were fixed.
You can take a look to previous blog posts.
```
We could also add something like that as there is middle windows information in 2.2.0-* blog posts & the windows blog post too.
